# Tensorflow 实战 Google 深度学习框架 | 学习笔记（二）

>Shoot on the moon and if you miss you will still be among the stars.

Caicloud Github :[tensorflow-tutorial](https://github.com/caicloud/tensorflow-tutorial)https://github.com/caicloud/tensorflow-tutorial

原 tutorial 使用的 Tensorflow 最新版本是 1.4.0 ，本人使用的 1.5.0 版本，所以部分代码会略有不同，该笔记仅为个人学习，理解使用。如有错误，还望批评指教。----ZJ

## 4 深层神经网络

### 4.1 深度学习与深层神经网络

wiki 深度学习定义：一类通过多层非线性变换对高复杂性数据建模算法的合集。

深度学习，两个重要特性：**多层 and 非线性。**

- 线性模型的局限性，线性模型能够解决的问题是有限的(解决线性可分的问题)。
    - 3.4.2 小节中，输出为所有输入的加权和，导致整个神经网络是一个线性模型。
    - （采用线性函数）任意层的全连接神经网络（FC）和单层神经网络模型的表达能力没有任何区别。
    - 线性模型可以解决线性可分的问题，如通过直线（或高维空间的平面）划分。
    - 现实世界中，绝大多数问题是无法线性分割的，所以出现 Relu 这样非线性的激活函数，来解决高复杂性的问题（深度学习的目的就是解决这样的问题）
    
- 激活函数实现去线性化
    - 如果每个神经元的输出通过一个非线性函数，那么整个神经网络就不再是线性的了。
    - 添加偏置项（bias）,偏置项 是神经网络中非常有用的一种结构。
    - 每个节点（神经元）的输出在 加权和的基础上，还做了一个非线性变换。
    - 常用非线性激活函数（tf.nn.relu, tf.sigmoid, tf.tanh）

- 多层网络解决异或运算
    - 在神经网络的发展史，有个很重要的问题，**异或问题**。
    - 1958 感知机模型，将输入进行加权和，没有隐藏层，然后经过激活函数最后得到输出。不能解决异或问题。
    - 异或问题（直观理解）：两个输入的符号相同时（同正，同负）输出为 0 ，否则（一正一负）输出为 1.
    - 加入隐藏层后，异或问题得到很好的解决
    - 深层网络有组合特征提取的功能，这个特性对于解决不易提取特征向量的问题（如图片识别，语音识别等）有很大的帮助。

### 4.2 损失函数的定义

**神经网络模型的效果以及优化的目标是通过损失函数（loss function）来定义的。**

监督学习的两大种类：分类问题 and 回归问题。

- 经典损失函数

- 自定义损失函数


如何判断一个输出向量和期望的向量有多接近？------交叉熵（cross entropy）是最常用的评判方法之一。

交叉熵 （cross entropy）:

给定两个概率分布 p,q 

$$H(p,q)= - \sum_{x}p(x)logq(x)$$

注意：交叉熵刻画的是两个概率之间的分布，但是神经网络的输出不一定是一个概率分布。

- 概率分布刻画了不同事件发生的概率，当事件总数有限的情况下，概率分布函数 $p(X=x)$ 满足：

$$\forall x \space\space p(X=x)\in[0,1]  且 \sum_{x}p(X=x)=1$$

- 任意事件发生的概率都在 0 和 1 之间
- 且总有某一个事件发生（概率的和为 1）

如何将神经网络前向传播的结果变成概率分布？

- softmax 回归是一个常用的方法

【图】4_10

假设原始的神经网络输出为 $y_1,y_2,\dots,y_n$,那么经过 softmax 处理之后的输出为：

$$ softmax(y)_i=y_i^{'}=\frac{e^{yi}}{\sum_{j=1}^{n}e^{yj}}$$

- 原始神经网络的输出被用作置信度来生成新的输出,新的输出满足概率分布的所有要求
- 新的输出可以理解为经过神经网络的推导，一个样本为不同类别的概率分别有多大
- 因此神经网络的输出变为了一个概率分布，从而可以通过交叉熵来计算预测概率分布和真实答案的概率分布之间的距离。
- 交叉熵函数不是对称的 （$H(p.q)\neq H(q,p)$）,刻画的是通过概率 q 来表达概率分布 p 的困难程度
- 在神经网络中，p 代表正确标签，q 代表预测值

In [2]:
import tensorflow as tf

tf.__version__

'1.5.0'

In [ ]:
'''
Tensorflow 实现交叉熵

'''

cross_entropy = -tf.reduce_mean(y_* tf.log(tf.clip_by_value(y, 1e-10, 1.0)))



reduce_mean 可以理解为 $ \frac{1}{n} \sum$ ,n 代表一个 batch 中样例的数量

其中：`y_ `代表正确结果， `y` 代表预测值

上面一行代码，包含了 4 种不同的 Tensorflow 运算。

- `tf.clip_by_value`函数可以将在一个张量中的数值限制在一个范围之内，可以避免一些运算错误，（如 log0 是无效的） 

  - 小于 2.5 的数值替换为 2.5 ，大于 4.5 的数值 替换为 4.5
  - 通过 `tf.clip_by_value`  函数可以保证在进行 log 运算时，不会出现 log0 这样的错误

In [3]:
import tensorflow as tf

v = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])

with tf.Session() as sess:
    print(tf.clip_by_value(v, 2.5, 4.5).eval())


[[2.5 2.5 3. ]
 [4.  4.5 4.5]]


- 第二个运算是 `tf.log` 函数，完成对张量中所有函数一次求对数的功能

In [5]:
v = tf.constant([1.0, 2.0, 3.0])

with tf.Session() as sess:
    
    print(tf.log(v).eval())

[0.        0.6931472 1.0986123]


- 第三个是乘法运算，在实现交叉熵的代码中直接将两个矩阵通过 `*`操作相乘，这个操作不是矩阵乘法，而是元素之间直接相乘 （element-wise）
- 矩阵乘法需要使用 `tf.matmul` 函数完成。下面示例给出两者区别：

In [8]:
v1 = tf.constant([[1.0, 2.0], [3.0, 4.0]])
v2 = tf.constant([[5.0, 6.0], [7.0, 8.0]])

with tf.Session() as sess:
    print('逐元乘积：\n',(v1*v2).eval())
    print('矩阵乘法:\n', tf.matmul(v1, v2).eval())

逐元乘积：
 [[ 5. 12.]
 [21. 32.]]
矩阵乘法:
 [[19. 22.]
 [43. 50.]]


- `v1 * v2` 的结果是每个位置上对应元素的乘积，如结果中（1,1）元素值 5 是 `1×5 = 5`
- `tf.matmul` 函数完成矩阵乘法 ，如 （1,1）元素值 19 是 `1× 5 + 2× 7 = 19`

- 上面三个运算完成了对每一个样例中每个类别交叉熵 `p(x)logq(x)`的计算
- 这三步计算分别得到的结果是一个 `n × m`的二维矩阵，n 为一个 batch 中样例的数量， m 为分类的类别数量。
- 根据该交叉熵的公式，应该将每行中的 m 个结果相加得到所有样例的交叉熵
- 然后再对这 n 行取平均得到一个 batch 的平均交叉熵
- 因为分类问题的类别数量是不变的，所以可以直接对整个矩阵做平均而不改变计算结果的意义
- 简单展示 `tf.reduce_mean` 函数的使用方法


In [9]:
v = tf.constant([[1.0,2.0,3.0],[4.0,5.0,6.0]])

with tf.Session() as sess:
    print(tf.reduce_mean(v).eval())

3.5


- 因为 交叉熵一般会和 softmax 回归一起使用，所以 Tensorflow 对这两个功能进行了统一封装 ，并提供 `tf.nn.softmax_cross_entopy_with_logits` 函数
- 可以通过下面代码来实现使用了 softmax 回归之后的交叉熵损失函数

In [ ]:
# y 是预测输出值，y_ 是标准答案
cross_entopy = tf.nn.softmax_cross_entropy_with_logits(y,y_)


- 回归问题解决的是对具体数值的预测，如房价预测，销量预测等。
- 解决回归问题的神经网络一般只有一个输出节点，输出值为预测值。
- 回归问题最常用的是损失函数是均方误差（MES, mean squared error）

$$ MSE(y,y^{'}) = \frac{\sum_{i=1}^{n}(y_i-y^{'})^2}{n}$$

- $y_i$ 为一个 batch 中第 i 个数据的正确答案，$y^{'}$ 为神经网络给出的预测值
- Tensorflow 实现均方误差损失函数

```
mse = tf.reduce_mean(tf.square(y_ - y))

```

其中：`y_ `代表标准答案， `y` 代表神经网络输出值

### 4.2.2 自定义损失函数

- Tensorflow 支持任意优化的自定义损失函数
- 下面公式给出一个 当预测多于真实值和预测少于真实值时有不同系数的损失函数

$$ Loss(y, y^{'}) = \sum_{i=1}^{n}f(y_i,y_i^{'}), \space\space  f(x,y)=\left\{
\begin{aligned}
a(x-y) & & x> y \\
b(y-x) & & x≤y \\
\end{aligned}\right.  $$

- $y_i$ 为一个 batch 中第 i 个数据的正确答案，$y^{'}$ 为神经网络给出的预测值 a,b 是常量，如 a = 10 , b = 1,TF 实现：


In [ ]:
loss = tf.reduce_sum(tf.where(tf.greater(v1,v2),(v1 - v2) * a, (v2- v1) *b ))

- 用到了 `tf.greater` 和 `tf.where` 来实现选择操作
- `tf.greater` 的输入是两个张量，此函数会比较这两个输入张量中每一个元素的大小，并返回比较结果
- 当 `tf.greater `  的输入张量维度不一样是，TF 会进行类似 Numpy 广播（broadcasting） 的处理。
- `tf.where` 有三个参数，1.选择条件根据，当选择条件 为 True 时，`tf.where` 函数会选择第二个参数中的值，否则使用第三个参数中的值。
- 注意：`tf.where` 函数判断和选择都是在元素级别进行，下面展示用法：

In [12]:
v1 = tf.constant([1.0, 2.0, 3.0, 4.0])
v2 = tf.constant([4.0, 3.0, 2.0, 1.0])

sess = tf.InteractiveSession()

print(tf.greater(v1, v2).eval())

print(tf.where(tf.greater(v1,v2),v1,v2).eval())

sess.close()

[False False  True  True]
[4. 3. 3. 4.]


下面通过一个简单的神经网络程序来讲解损失函数对模型训练结果的影响。

- 两个输入节点，一个输出节点，没有隐藏层 

In [11]:
import tensorflow as tf
from numpy.random import RandomState

'''
1. 定义神经网络的相关参数和变量。

'''

batch_size = 8

# 输入两个节点
x = tf.placeholder(tf.float32, shape=(None, 2), name='x-input')

# 回归问题一般只有一个输出节点
y_ = tf.placeholder(tf.float32, shape=(None, 1), name='y-input')

# 定义一个单层神经网络的前向传播过程，这里是简单加权和 stddev=1 标准差
w1 = tf.Variable(tf.random_normal([2, 1], stddev=1, seed=1))
y = tf.matmul(x, w1)


'''
2. 设置自定义的损失函数。
 
'''
# 定义预测多了和预测少了的 成本
loss_less = 10 # 预测少了 会损失 10块
loss_more = 1  # 预测多了 只损失 1 块
loss = tf.reduce_sum(tf.where(tf.greater(y,y_),
                             (y - y_) * loss_more,
                             (y_ - y) * loss_less))

train_step = tf.train.AdamOptimizer(0.001).minimize(loss)

'''
3. 生成模拟数据集。
'''

# 通过随机数生成一个数据模拟集 128 个样本，每个样本两个特征
rdm = RandomState(1)

X = rdm.rand(128, 2)

print(X.shape)
# print(X)
'''
设置回归的正确值为两个输入的和加上一个随机量，之所以要加上一个随机量是为了
加入不可预测的噪音，否则不同损失函数的意义就不大了，因为不同损失函数都会在能
完全预测正确的时候最低，一般来说噪音是一个均值为 0 的小量，所以这里的
噪音设置为 -0.05 ~ 0.05 的随机数

'''
# Y 正确值 是一个 list (x1, x2) in X 一个样本有两个特征，128 行 2 列

Y = [[x1 + x2 + (rdm.rand()/10.0-0.05)]  for (x1, x2) in X]

'''
4.训练模型。

'''        

with tf.Session() as sess:
    init_op = tf.global_variables_initializer()
    sess.run(init_op)
    STEPS = 5000
    for i in range(STEPS):
        start = (i*batch_size) % 128
        end = (i*batch_size) % 128 + batch_size
        sess.run(train_step, feed_dict={x: X[start:end], y_: Y[start:end]})
        if i % 1000 == 0:
            print("After %d training step(s), w1 is: " % (i))
            print(sess.run(w1), "\n") 
    print("Final w1 is: \n", sess.run(w1)) 

(128, 2)
After 0 training step(s), w1 is: 
[[-0.81031823]
 [ 1.4855988 ]] 

After 1000 training step(s), w1 is: 
[[0.01247112]
 [2.1385448 ]] 

After 2000 training step(s), w1 is: 
[[0.45567414]
 [2.1706066 ]] 

After 3000 training step(s), w1 is: 
[[0.69968724]
 [1.8465308 ]] 

After 4000 training step(s), w1 is: 
[[0.89886665]
 [1.2973602 ]] 

Final w1 is: 
 [[1.019347 ]
 [1.0428089]]


In [17]:
'''
5. 重新定义损失函数，使得预测多了的损失大，于是模型应该偏向少的方向预测。

'''
loss_less = 1
loss_more = 10 # 预测多了的损失大

loss = tf.reduce_sum(tf.where(tf.greater(y, y_),(y - y_)* loss_more, (y_ - y) * loss_less))
train_step = tf.train.AdamOptimizer(0.001).minimize(loss)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    STEPS = 5000
    for i in range(STEPS):
        start = (i*batch_size) % 128
        end = (i*batch_size) % 128 + batch_size
        sess.run(train_step, feed_dict={x:X[start:end], y_:Y[start:end]})
        if i % 1000 == 0:
            print("After %d training step(s), w1 is: " % (i))
            print(sess.run(w1), "\n") 
    
        print("Final w1 is: \n", sess.run(w1))


After 0 training step(s), w1 is: 
[[-0.8123182]
 [ 1.4835987]] 

Final w1 is: 
 [[-0.8123182]
 [ 1.4835987]]
Final w1 is: 
 [[-0.81303465]
 [ 1.4826417 ]]
Final w1 is: 
 [[-0.81253535]
 [ 1.4817047 ]]
Final w1 is: 
 [[-0.8119145]
 [ 1.4807667]]
Final w1 is: 
 [[-0.8111701]
 [ 1.4798151]]
Final w1 is: 
 [[-0.81035256]
 [ 1.4789057 ]]
Final w1 is: 
 [[-0.8094773]
 [ 1.4779707]]
Final w1 is: 
 [[-0.808571 ]
 [ 1.4770529]]
Final w1 is: 
 [[-0.8076296]
 [ 1.4762905]]
Final w1 is: 
 [[-0.8066966]
 [ 1.4755214]]
Final w1 is: 
 [[-0.80576235]
 [ 1.4748191 ]]
Final w1 is: 
 [[-0.8048966]
 [ 1.474087 ]]
Final w1 is: 
 [[-0.8039802]
 [ 1.4734688]]
Final w1 is: 
 [[-0.8031049]
 [ 1.472757 ]]
Final w1 is: 
 [[-0.80219334]
 [ 1.4720993 ]]
Final w1 is: 
 [[-0.8012718]
 [ 1.4713938]]
Final w1 is: 
 [[-0.80047035]
 [ 1.4706074 ]]
Final w1 is: 
 [[-0.79974467]
 [ 1.469795  ]]
Final w1 is: 
 [[-0.79897094]
 [ 1.4689691 ]]
Final w1 is: 
 [[-0.79820436]
 [ 1.4681253 ]]
Final w1 is: 
 [[-0.79741085]
 [ 1.46

 [[-0.53402334]
 [ 1.2758325 ]]
Final w1 is: 
 [[-0.5330304]
 [ 1.2752793]]
Final w1 is: 
 [[-0.53201187]
 [ 1.2747679 ]]
Final w1 is: 
 [[-0.53100216]
 [ 1.2741842 ]]
Final w1 is: 
 [[-0.53004247]
 [ 1.2735748 ]]
Final w1 is: 
 [[-0.5290823]
 [ 1.2730861]]
Final w1 is: 
 [[-0.52807313]
 [ 1.2725995 ]]
Final w1 is: 
 [[-0.5270732]
 [ 1.2721003]]
Final w1 is: 
 [[-0.5260836]
 [ 1.2715098]]
Final w1 is: 
 [[-0.52510285]
 [ 1.2709098 ]]
Final w1 is: 
 [[-0.5240984]
 [ 1.2702125]]
Final w1 is: 
 [[-0.5230985]
 [ 1.2694957]]
Final w1 is: 
 [[-0.52205634]
 [ 1.2688872 ]]
Final w1 is: 
 [[-0.52104807]
 [ 1.2682551 ]]
Final w1 is: 
 [[-0.52006143]
 [ 1.2676743 ]]
Final w1 is: 
 [[-0.5190626]
 [ 1.2671945]]
Final w1 is: 
 [[-0.51800203]
 [ 1.2667952 ]]
Final w1 is: 
 [[-0.5170094]
 [ 1.2662462]]
Final w1 is: 
 [[-0.5159915]
 [ 1.2657386]]
Final w1 is: 
 [[-0.5149827]
 [ 1.2651579]]
Final w1 is: 
 [[-0.5140241]
 [ 1.2645509]]
Final w1 is: 
 [[-0.51306516]
 [ 1.2640646 ]]
Final w1 is: 
 [[-0.5120

 [ 1.179712]]
Final w1 is: 
 [[-0.2979762]
 [ 1.1793075]]
Final w1 is: 
 [[-0.29698887]
 [ 1.1790103 ]]
Final w1 is: 
 [[-0.29596257]
 [ 1.1786914 ]]
Final w1 is: 
 [[-0.2949523]
 [ 1.1783366]]
Final w1 is: 
 [[-0.29391173]
 [ 1.177987  ]]
Final w1 is: 
 [[-0.2928518]
 [ 1.1777478]]
Final w1 is: 
 [[-0.29178268]
 [ 1.177356  ]]
Final w1 is: 
 [[-0.29073006]
 [ 1.176904  ]]
Final w1 is: 
 [[-0.28964764]
 [ 1.1765387 ]]
Final w1 is: 
 [[-0.28860733]
 [ 1.1761162 ]]
Final w1 is: 
 [[-0.28757292]
 [ 1.1758012 ]]
Final w1 is: 
 [[-0.28653714]
 [ 1.1755658 ]]
Final w1 is: 
 [[-0.28545126]
 [ 1.1753907 ]]
Final w1 is: 
 [[-0.28439406]
 [ 1.1751487 ]]
Final w1 is: 
 [[-0.2833246]
 [ 1.1749165]]
Final w1 is: 
 [[-0.28227457]
 [ 1.1745689 ]]
Final w1 is: 
 [[-0.281282]
 [ 1.174162]]
Final w1 is: 
 [[-0.2802979]
 [ 1.1738633]]
Final w1 is: 
 [[-0.2792749]
 [ 1.1735426]]
Final w1 is: 
 [[-0.27826792]
 [ 1.173186  ]]
Final w1 is: 
 [[-0.27723065]
 [ 1.1728344 ]]
Final w1 is: 
 [[-0.2761741]
 [ 1.17

 [[-0.02449179]
 [ 1.0984439 ]]
Final w1 is: 
 [[-0.02345964]
 [ 1.0984147 ]]
Final w1 is: 
 [[-0.02237755]
 [ 1.0984674 ]]
Final w1 is: 
 [[-0.02132664]
 [ 1.0984251 ]]
Final w1 is: 
 [[-0.02026687]
 [ 1.0983721 ]]
Final w1 is: 
 [[-0.01922868]
 [ 1.0981771 ]]
Final w1 is: 
 [[-0.01824863]
 [ 1.0979018 ]]
Final w1 is: 
 [[-0.01727905]
 [ 1.0977257 ]]
Final w1 is: 
 [[-0.01627366]
 [ 1.0975125 ]]
Final w1 is: 
 [[-0.01528559]
 [ 1.0972484 ]]
Final w1 is: 
 [[-0.01425358]
 [ 1.0970697 ]]
Final w1 is: 
 [[-0.01320589]
 [ 1.0969895 ]]
Final w1 is: 
 [[-0.01215199]
 [ 1.0967286 ]]
Final w1 is: 
 [[-0.01111649]
 [ 1.0963879 ]]
Final w1 is: 
 [[-0.01005427]
 [ 1.0961257 ]]
Final w1 is: 
 [[-0.00901132]
 [ 1.0959541 ]]
Final w1 is: 
 [[-0.00797822]
 [ 1.0958692 ]]
Final w1 is: 
 [[-0.00694753]
 [ 1.0958445 ]]
Final w1 is: 
 [[-0.00586704]
 [ 1.0959015 ]]
Final w1 is: 
 [[-0.00481772]
 [ 1.0958629 ]]
Final w1 is: 
 [[-0.00375962]
 [ 1.095813  ]]
Final w1 is: 
 [[-0.0027231]
 [ 1.0956205]]
Fina

Final w1 is: 
 [[0.42956868]
 [1.0623162 ]]
Final w1 is: 
 [[0.43057027]
 [1.0623088 ]]
Final w1 is: 
 [[0.4315692]
 [1.0623568]]
Final w1 is: 
 [[0.4326162]
 [1.0624835]]
Final w1 is: 
 [[0.4336395]
 [1.062673 ]]
Final w1 is: 
 [[0.43467042]
 [1.0628277 ]]
Final w1 is: 
 [[0.43567958]
 [1.0628109 ]]
Final w1 is: 
 [[0.4366317]
 [1.0626903]]
Final w1 is: 
 [[0.437573 ]
 [1.0626576]]
Final w1 is: 
 [[0.43854833]
 [1.0625703 ]]
Final w1 is: 
 [[0.43951243]
 [1.0625376 ]]
Final w1 is: 
 [[0.44051802]
 [1.0625705 ]]
Final w1 is: 
 [[0.44153786]
 [1.0626854 ]]
Final w1 is: 
 [[0.44256282]
 [1.0625888 ]]
Final w1 is: 
 [[0.44356918]
 [1.06239   ]]
Final w1 is: 
 [[0.4446008]
 [1.0622581]]
Final w1 is: 
 [[0.4456131]
 [1.0622076]]
Final w1 is: 
 [[0.44661528]
 [1.0622357 ]]
Final w1 is: 
 [[0.44761464]
 [1.0623157 ]]
Final w1 is: 
 [[0.44866195]
 [1.0624714 ]]
Final w1 is: 
 [[0.44968545]
 [1.062687  ]]
Final w1 is: 
 [[0.45071647]
 [1.0628654 ]]
Final w1 is: 
 [[0.45172563]
 [1.0628695 ]]
Fi

Final w1 is: 
 [[0.7809162]
 [1.0317718]]
Final w1 is: 
 [[0.78185123]
 [1.0316994 ]]
Final w1 is: 
 [[0.78272897]
 [1.0315727 ]]
Final w1 is: 
 [[0.7836673]
 [1.0315442]]
Final w1 is: 
 [[0.7845929]
 [1.0315958]]
Final w1 is: 
 [[0.7855361]
 [1.031626 ]]
Final w1 is: 
 [[0.7864664]
 [1.0314934]]
Final w1 is: 
 [[0.7873563]
 [1.0314448]]
Final w1 is: 
 [[0.7881563]
 [1.0312401]]
Final w1 is: 
 [[0.7889722]
 [1.0309147]]
Final w1 is: 
 [[0.789793 ]
 [1.0306685]]
Final w1 is: 
 [[0.7906701]
 [1.0305107]]
Final w1 is: 
 [[0.7915746]
 [1.0304558]]
Final w1 is: 
 [[0.792496 ]
 [1.0302023]]
Final w1 is: 
 [[0.79340935]
 [1.0298597 ]]
Final w1 is: 
 [[0.7943576]
 [1.0295994]]
Final w1 is: 
 [[0.79529506]
 [1.0294348 ]]
Final w1 is: 
 [[0.79623014]
 [1.0293618 ]]
Final w1 is: 
 [[0.79710793]
 [1.0292346 ]]
Final w1 is: 
 [[0.79804647]
 [1.0292057 ]]
Final w1 is: 
 [[0.79897225]
 [1.0292567 ]]
Final w1 is: 
 [[0.7999157]
 [1.0292864]]
Final w1 is: 
 [[0.8008462]
 [1.0291535]]
Final w1 is: 
 [[0

 [0.9813667]]
Final w1 is: 
 [[0.95353734]
 [0.9813056 ]]
Final w1 is: 
 [[0.9538224 ]
 [0.98131627]]
Final w1 is: 
 [[0.9539559 ]
 [0.98134965]]
Final w1 is: 
 [[0.9541829 ]
 [0.98128575]]
Final w1 is: 
 [[0.95447  ]
 [0.9811102]]
Final w1 is: 
 [[0.95450115]
 [0.9809726 ]]
Final w1 is: 
 [[0.9544358 ]
 [0.98091817]]
Final w1 is: 
 [[0.95446676]
 [0.9809468 ]]
Final w1 is: 
 [[0.95431966]
 [0.9808589 ]]
Final w1 is: 
 [[0.9543337]
 [0.980868 ]]
Final w1 is: 
 [[0.9544259 ]
 [0.98095566]]
Final w1 is: 
 [[0.9544729]
 [0.9809144]]
Final w1 is: 
 [[0.9545952 ]
 [0.98089767]]
Final w1 is: 
 [[0.9547106]
 [0.980938 ]]
Final w1 is: 
 [[0.95479566]
 [0.9809297 ]]
Final w1 is: 
 [[0.9549722 ]
 [0.98091847]]
Final w1 is: 
 [[0.95521605]
 [0.98095673]]
Final w1 is: 
 [[0.9555719]
 [0.9810571]]
Final w1 is: 
 [[0.9557695 ]
 [0.98117137]]
Final w1 is: 
 [[0.95605385]
 [0.9811799 ]]
Final w1 is: 
 [[0.95639235]
 [0.981069  ]]
Final w1 is: 
 [[0.95622104]
 [0.98094547]]
Final w1 is: 
 [[0.95597464]

Final w1 is: 
 [[0.9557112]
 [0.9813241]]
Final w1 is: 
 [[0.95587915]
 [0.981448  ]]
Final w1 is: 
 [[0.9561268 ]
 [0.98145944]]
Final w1 is: 
 [[0.9564244]
 [0.9813439]]
Final w1 is: 
 [[0.9562611]
 [0.9812147]]
Final w1 is: 
 [[0.95603037]
 [0.9811724 ]]
Final w1 is: 
 [[0.9559031]
 [0.981217 ]]
Final w1 is: 
 [[0.95563185]
 [0.98113614]]
Final w1 is: 
 [[0.95551914]
 [0.98115724]]
Final w1 is: 
 [[0.9554889 ]
 [0.98126096]]
Final w1 is: 
 [[0.95542943]
 [0.98122627]]
Final w1 is: 
 [[0.9554477]
 [0.9812168]]
Final w1 is: 
 [[0.9554688]
 [0.9812673]]
Final w1 is: 
 [[0.95547086]
 [0.9812652 ]]
Final w1 is: 
 [[0.9555625 ]
 [0.98125935]]
Final w1 is: 
 [[0.9555825]
 [0.9812948]]
Final w1 is: 
 [[0.955723]
 [0.981397]]
Final w1 is: 
 [[0.9557394]
 [0.9815144]]
Final w1 is: 
 [[0.9558498 ]
 [0.98151964]]
Final w1 is: 
 [[0.9560233]
 [0.9813981]]
Final w1 is: 
 [[0.9557528 ]
 [0.98126334]]
Final w1 is: 
 [[0.9554263]
 [0.9812163]]
Final w1 is: 
 [[0.9552122]
 [0.981257 ]]
Final w1 is: 


 [[0.9547262]
 [0.9813157]]
Final w1 is: 
 [[0.9547979]
 [0.9813103]]
Final w1 is: 
 [[0.9549334]
 [0.9813596]]
Final w1 is: 
 [[0.9551693 ]
 [0.98147786]]
Final w1 is: 
 [[0.95527923]
 [0.9816111 ]]
Final w1 is: 
 [[0.9554672]
 [0.9816253]]
Final w1 is: 
 [[0.9557054 ]
 [0.98150545]]
Final w1 is: 
 [[0.95552164]
 [0.98137087]]
Final w1 is: 
 [[0.95527875]
 [0.98132783]]
Final w1 is: 
 [[0.9551345 ]
 [0.98137635]]
Final w1 is: 
 [[0.9548596 ]
 [0.98129237]]
Final w1 is: 
 [[0.95473385]
 [0.9813158 ]]
Final w1 is: 
 [[0.9546866 ]
 [0.98142624]]
Final w1 is: 
 [[0.95461416]
 [0.9813906 ]]
Final w1 is: 
 [[0.9546154]
 [0.9813814]]
Final w1 is: 
 [[0.9546209]
 [0.9814354]]
Final w1 is: 
 [[0.95461017]
 [0.98143387]]
Final w1 is: 
 [[0.95468354]
 [0.9814283 ]]
Final w1 is: 
 [[0.9548201]
 [0.9814777]]
Final w1 is: 
 [[0.95505637]
 [0.9815962 ]]
Final w1 is: 
 [[0.9551672]
 [0.9817296]]
Final w1 is: 
 [[0.95535547]
 [0.9817439 ]]
Final w1 is: 
 [[0.9555935]
 [0.9816236]]
Final w1 is: 
 [[0.9

 [[0.9560561 ]
 [0.98078424]]
Final w1 is: 
 [[0.95612013]
 [0.98082846]]
Final w1 is: 
 [[0.9562875 ]
 [0.98094547]]
Final w1 is: 
 [[0.95633954]
 [0.98107874]]
Final w1 is: 
 [[0.95647216]
 [0.98108834]]
Final w1 is: 
 [[0.95665795]
 [0.98095834]]
Final w1 is: 
 [[0.9564419]
 [0.9808135]]
Final w1 is: 
 [[0.95617276]
 [0.9807647 ]]
Final w1 is: 
 [[0.95600224]
 [0.9808119 ]]
Final w1 is: 
 [[0.9557089 ]
 [0.98072106]]
Final w1 is: 
 [[0.9555621 ]
 [0.98074275]]
Final w1 is: 
 [[0.95549357]
 [0.9808555 ]]
Final w1 is: 
 [[0.9554031]
 [0.980816 ]]
Final w1 is: 
 [[0.9553856]
 [0.9808044]]
Final w1 is: 
 [[0.9553741]
 [0.9808589]]
Final w1 is: 
 [[0.9553487 ]
 [0.98085564]]
Final w1 is: 
 [[0.9554058]
 [0.9808484]]
Final w1 is: 
 [[0.9554015 ]
 [0.98088664]]
Final w1 is: 
 [[0.95550686]
 [0.98099846]]
Final w1 is: 
 [[0.9555035]
 [0.9811271]]
Final w1 is: 
 [[0.95558596]
 [0.98113227]]
Final w1 is: 
 [[0.95572627]
 [0.98099804]]
Final w1 is: 
 [[0.9556711]
 [0.9809012]]
Final w1 is: 
 [

 [0.9807907]]
Final w1 is: 
 [[0.95591927]
 [0.9808714 ]]
Final w1 is: 
 [[0.95598286]
 [0.9808898 ]]
Final w1 is: 
 [[0.95611876]
 [0.9809019 ]]
Final w1 is: 
 [[0.95618635]
 [0.9809592 ]]
Final w1 is: 
 [[0.95635456]
 [0.98109084]]
Final w1 is: 
 [[0.95640945]
 [0.98123837]]
Final w1 is: 
 [[0.9565428]
 [0.9812566]]
Final w1 is: 
 [[0.9567278 ]
 [0.98112917]]
Final w1 is: 
 [[0.95651925]
 [0.9809856 ]]
Final w1 is: 
 [[0.9562584]
 [0.980941 ]]
Final w1 is: 
 [[0.95609385]
 [0.9809954 ]]
Final w1 is: 
 [[0.9558088 ]
 [0.98090607]]
Final w1 is: 
 [[0.955667  ]
 [0.98093295]]
Final w1 is: 
 [[0.95560163]
 [0.98105395]]
Final w1 is: 
 [[0.95551455]
 [0.9810166 ]]
Final w1 is: 
 [[0.9554989 ]
 [0.98100775]]
Final w1 is: 
 [[0.95548886]
 [0.98106724]]
Final w1 is: 
 [[0.9554651 ]
 [0.98106647]]
Final w1 is: 
 [[0.955522 ]
 [0.9810613]]
Final w1 is: 
 [[0.95551866]
 [0.9811031 ]]
Final w1 is: 
 [[0.9556227 ]
 [0.98122096]]
Final w1 is: 
 [[0.95562017]
 [0.98135614]]
Final w1 is: 
 [[0.95570

 [0.98010206]]
Final w1 is: 
 [[0.9555055]
 [0.9800852]]
Final w1 is: 
 [[0.9555659]
 [0.9801391]]
Final w1 is: 
 [[0.9556059]
 [0.9801319]]
Final w1 is: 
 [[0.95571864]
 [0.9801209 ]]
Final w1 is: 
 [[0.9557667]
 [0.9801586]]
Final w1 is: 
 [[0.9559147 ]
 [0.98027474]]
Final w1 is: 
 [[0.9559538]
 [0.980409 ]]
Final w1 is: 
 [[0.9560746 ]
 [0.98063236]]
Final w1 is: 
 [[0.95624673]
 [0.9806856 ]]
Final w1 is: 
 [[0.9560358 ]
 [0.98070395]]
Final w1 is: 
 [[0.9557746 ]
 [0.98080736]]
Final w1 is: 
 [[0.955608 ]
 [0.9809975]]
Final w1 is: 
 [[0.9553245 ]
 [0.98102653]]
Final w1 is: 
 [[0.95518124]
 [0.9811628 ]]
Final w1 is: 
 [[0.95511305]
 [0.98138475]]
Final w1 is: 
 [[0.9550242 ]
 [0.98143435]]
Final w1 is: 
 [[0.9550053]
 [0.9815045]]
Final w1 is: 
 [[0.9549923]
 [0.9816368]]
Final w1 is: 
 [[0.9549662 ]
 [0.98170024]]
Final w1 is: 
 [[0.9550192]
 [0.9817527]]
Final w1 is: 
 [[0.95513177]
 [0.9818603 ]]
Final w1 is: 
 [[0.9553375]
 [0.9820394]]
Final w1 is: 
 [[0.9554289]
 [0.98223

 [[0.95622903]
 [0.9810949 ]]
Final w1 is: 
 [[0.9564012]
 [0.9809537]]
Final w1 is: 
 [[0.95620084]
 [0.9807964 ]]
Final w1 is: 
 [[0.95595115]
 [0.9807434 ]]
Final w1 is: 
 [[0.95579296]
 [0.98079467]]
Final w1 is: 
 [[0.9555208]
 [0.980696 ]]
Final w1 is: 
 [[0.9553846 ]
 [0.98071957]]
Final w1 is: 
 [[0.955321]
 [0.980842]]
Final w1 is: 
 [[0.95523703]
 [0.98079914]]
Final w1 is: 
 [[0.9552209]
 [0.9807865]]
Final w1 is: 
 [[0.9552102]
 [0.9808457]]
Final w1 is: 
 [[0.9551866]
 [0.9808422]]
Final w1 is: 
 [[0.95523965]
 [0.9808343 ]]
Final w1 is: 
 [[0.95535046]
 [0.9808887 ]]
Final w1 is: 
 [[0.9555516 ]
 [0.98102164]]
Final w1 is: 
 [[0.95564157]
 [0.98117167]]
Final w1 is: 
 [[0.9558018]
 [0.9811867]]
Final w1 is: 
 [[0.9560074 ]
 [0.98104954]]
Final w1 is: 
 [[0.95583785]
 [0.9808958 ]]
Final w1 is: 
 [[0.9556161]
 [0.9808461]]
Final w1 is: 
 [[0.95548284]
 [0.98090047]]
Final w1 is: 
 [[0.95523345]
 [0.98080444]]
Final w1 is: 
 [[0.95511746]
 [0.98083043]]
Final w1 is: 
 [[0.9

Final w1 is: 
 [[0.9554687]
 [0.9819434]]
Final w1 is: 
 [[0.9555154 ]
 [0.98212874]]
Final w1 is: 
 [[0.9556361 ]
 [0.98217386]]
Final w1 is: 
 [[0.9558056]
 [0.9820616]]
Final w1 is: 
 [[0.9556065 ]
 [0.98192996]]
Final w1 is: 
 [[0.95535874]
 [0.98190135]]
Final w1 is: 
 [[0.95520157]
 [0.98197603]]
Final w1 is: 
 [[0.9549317]
 [0.9818963]]
Final w1 is: 
 [[0.9547964]
 [0.9819385]]
Final w1 is: 
 [[0.9546419]
 [0.9818435]]
Final w1 is: 
 [[0.95447636]
 [0.98160285]]
Final w1 is: 
 [[0.9543861 ]
 [0.98141253]]
Final w1 is: 
 [[0.95430875]
 [0.9813128 ]]
Final w1 is: 
 [[0.9542252 ]
 [0.98116535]]
Final w1 is: 
 [[0.9542235]
 [0.9810278]]
Final w1 is: 
 [[0.9542844]
 [0.9809664]]
Final w1 is: 
 [[0.9544397]
 [0.9809962]]
Final w1 is: 
 [[0.95448923]
 [0.9810538 ]]
Final w1 is: 
 [[0.95461226]
 [0.98098403]]
Final w1 is: 
 [[0.95478374]
 [0.98076844]]
Final w1 is: 
 [[0.9547713]
 [0.9806007]]
Final w1 is: 
 [[0.95469135]
 [0.9805397 ]]
Final w1 is: 
 [[0.95468545]
 [0.9805852 ]]
Final 

 [[0.955934  ]
 [0.98118323]]
Final w1 is: 
 [[0.9561468]
 [0.981044 ]]
Final w1 is: 
 [[0.9559897]
 [0.9808877]]
Final w1 is: 
 [[0.95578045]
 [0.9808378 ]]
Final w1 is: 
 [[0.9556573]
 [0.9808944]]
Final w1 is: 
 [[0.9554192]
 [0.9807969]]
Final w1 is: 
 [[0.9553115 ]
 [0.98082423]]
Final w1 is: 
 [[0.9552724]
 [0.9809527]]
Final w1 is: 
 [[0.955211 ]
 [0.9809114]]
Final w1 is: 
 [[0.955214 ]
 [0.9809008]]
Final w1 is: 
 [[0.9552206 ]
 [0.98096365]]
Final w1 is: 
 [[0.9552127]
 [0.9809619]]
Final w1 is: 
 [[0.9552785 ]
 [0.98095554]]
Final w1 is: 
 [[0.9553996]
 [0.9810129]]
Final w1 is: 
 [[0.955608 ]
 [0.9811505]]
Final w1 is: 
 [[0.95570624]
 [0.98130554]]
Final w1 is: 
 [[0.9558724]
 [0.9813221]]
Final w1 is: 
 [[0.95608217]
 [0.9811826 ]]
Final w1 is: 
 [[0.9559231]
 [0.981026 ]]
Final w1 is: 
 [[0.95571214]
 [0.9809759 ]]
Final w1 is: 
 [[0.9555873]
 [0.9810324]]
Final w1 is: 
 [[0.955348 ]
 [0.9809347]]
Final w1 is: 
 [[0.95523906]
 [0.9809619 ]]
Final w1 is: 
 [[0.9551987]
 [

Final w1 is: 
 [[0.95486486]
 [0.98145497]]
Final w1 is: 
 [[0.9549387 ]
 [0.98152584]]
Final w1 is: 
 [[0.955009 ]
 [0.9816625]]
Final w1 is: 
 [[0.9550586]
 [0.9817269]]
Final w1 is: 
 [[0.95517594]
 [0.98178   ]]
Final w1 is: 
 [[0.9553433 ]
 [0.98189133]]
Final w1 is: 
 [[0.9555932]
 [0.9820782]]
Final w1 is: 
 [[0.955729 ]
 [0.9822778]]
Final w1 is: 
 [[0.9559288 ]
 [0.98233354]]
Final w1 is: 
 [[0.9561688 ]
 [0.98222804]]
Final w1 is: 
 [[0.95603734]
 [0.98210186]]
Final w1 is: 
 [[0.9558514 ]
 [0.98207986]]
Final w1 is: 
 [[0.955749  ]
 [0.98216236]]
Final w1 is: 
 [[0.95553005]
 [0.98208696]]
Final w1 is: 
 [[0.9554392]
 [0.9821351]]
Final w1 is: 
 [[0.9553252]
 [0.9820431]]
Final w1 is: 
 [[0.95519644]
 [0.98180234]]
Final w1 is: 
 [[0.95513856]
 [0.9816124 ]]
Final w1 is: 
 [[0.9550903]
 [0.9815142]]
Final w1 is: 
 [[0.9550331]
 [0.9813672]]
Final w1 is: 
 [[0.9550542]
 [0.9812299]]
Final w1 is: 
 [[0.95513487]
 [0.9811699 ]]
Final w1 is: 
 [[0.95530653]
 [0.9812025 ]]
Final 

In [19]:

'''
6. 定义损失函数为MSE。
'''
loss = tf.losses.mean_squared_error(y,y_)
train_step = tf.train.AdamOptimizer(0.001).minimize(loss)

with tf.Session() as sess:
    init_op = tf.global_variables_initializer()
    sess.run(init_op)
    STEPS = 5000
    for i in range(STEPS):
        start = (i*batch_size) % 128
        end = (i*batch_size) % 128 + batch_size
        sess.run(train_step, feed_dict={x: X[start:end], y_: Y[start:end]})
        if i % 1000 == 0:
            print("After %d training step(s), w1 is: " % (i))
            print(sess.run(w1), "\n") 
    
        print("Final w1 is: \n", sess.run(w1))

After 0 training step(s), w1 is: 
[[-0.81031823]
 [ 1.4855988 ]] 

Final w1 is: 
 [[-0.81031823]
 [ 1.4855988 ]]
Final w1 is: 
 [[-0.80941665]
 [ 1.4865069 ]]
Final w1 is: 
 [[-0.8085683]
 [ 1.4874406]]
Final w1 is: 
 [[-0.80770916]
 [ 1.4883332 ]]
Final w1 is: 
 [[-0.8068043]
 [ 1.4892628]]
Final w1 is: 
 [[-0.805891]
 [ 1.490211]]
Final w1 is: 
 [[-0.8049535]
 [ 1.4911743]]
Final w1 is: 
 [[-0.80400085]
 [ 1.4921538 ]]
Final w1 is: 
 [[-0.80303305]
 [ 1.493138  ]]
Final w1 is: 
 [[-0.80209213]
 [ 1.4940823 ]]
Final w1 is: 
 [[-0.8011821]
 [ 1.4950299]]
Final w1 is: 
 [[-0.8002777]
 [ 1.495945 ]]
Final w1 is: 
 [[-0.7993381]
 [ 1.4968998]]
Final w1 is: 
 [[-0.7984167]
 [ 1.4978073]]
Final w1 is: 
 [[-0.7974996]
 [ 1.4987341]]
Final w1 is: 
 [[-0.7965965]
 [ 1.4996561]]
Final w1 is: 
 [[-0.7957579]
 [ 1.5005246]]
Final w1 is: 
 [[-0.7949343]
 [ 1.5014063]]
Final w1 is: 
 [[-0.7940568]
 [ 1.5023218]]
Final w1 is: 
 [[-0.7931928]
 [ 1.503203 ]]
Final w1 is: 
 [[-0.7922794]
 [ 1.5041058]]

Final w1 is: 
 [[-0.53213996]
 [ 1.7314233 ]]
Final w1 is: 
 [[-0.5315182]
 [ 1.7318621]]
Final w1 is: 
 [[-0.5308287]
 [ 1.7323552]]
Final w1 is: 
 [[-0.53015083]
 [ 1.732829  ]]
Final w1 is: 
 [[-0.52940094]
 [ 1.7333441 ]]
Final w1 is: 
 [[-0.52865905]
 [ 1.7338524 ]]
Final w1 is: 
 [[-0.5278926]
 [ 1.7343698]]
Final w1 is: 
 [[-0.52711046]
 [ 1.7349185 ]]
Final w1 is: 
 [[-0.5262992]
 [ 1.7354971]]
Final w1 is: 
 [[-0.525518 ]
 [ 1.7360178]]
Final w1 is: 
 [[-0.524778 ]
 [ 1.7365323]]
Final w1 is: 
 [[-0.5240454]
 [ 1.7370125]]
Final w1 is: 
 [[-0.5232709]
 [ 1.7376066]]
Final w1 is: 
 [[-0.52251875]
 [ 1.7381191 ]]
Final w1 is: 
 [[-0.52178586]
 [ 1.7386354 ]]
Final w1 is: 
 [[-0.52107966]
 [ 1.7391086 ]]
Final w1 is: 
 [[-0.52043766]
 [ 1.7395184 ]]
Final w1 is: 
 [[-0.51982456]
 [ 1.7399338 ]]
Final w1 is: 
 [[-0.5191437]
 [ 1.7404042]]
Final w1 is: 
 [[-0.51847404]
 [ 1.7408566 ]]
Final w1 is: 
 [[-0.51773214]
 [ 1.7413514 ]]
Final w1 is: 
 [[-0.5169984]
 [ 1.7418386]]
Final w1

Final w1 is: 
 [[-0.30259418]
 [ 1.8331532 ]]
Final w1 is: 
 [[-0.30209896]
 [ 1.8331071 ]]
Final w1 is: 
 [[-0.30163538]
 [ 1.8330637 ]]
Final w1 is: 
 [[-0.3011049]
 [ 1.8330865]]
Final w1 is: 
 [[-0.30057654]
 [ 1.8331183 ]]
Final w1 is: 
 [[-0.2999731]
 [ 1.8332115]]
Final w1 is: 
 [[-0.29938093]
 [ 1.8332832 ]]
Final w1 is: 
 [[-0.2987652]
 [ 1.8333608]]
Final w1 is: 
 [[-0.29813614]
 [ 1.8334761 ]]
Final w1 is: 
 [[-0.29747435]
 [ 1.8336469 ]]
Final w1 is: 
 [[-0.2968393]
 [ 1.83375  ]]
Final w1 is: 
 [[-0.29624608]
 [ 1.833845  ]]
Final w1 is: 
 [[-0.2956537]
 [ 1.8339145]]
Final w1 is: 
 [[-0.2950314]
 [ 1.834104 ]]
Final w1 is: 
 [[-0.2944275]
 [ 1.8341893]]
Final w1 is: 
 [[-0.29385  ]
 [ 1.8342597]]
Final w1 is: 
 [[-0.29330242]
 [ 1.8342559 ]]
Final w1 is: 
 [[-0.2928125]
 [ 1.8341914]]
Final w1 is: 
 [[-0.29235432]
 [ 1.8341297 ]]
Final w1 is: 
 [[-0.29182926]
 [ 1.8341345 ]]
Final w1 is: 
 [[-0.29130596]
 [ 1.8341494 ]]
Final w1 is: 
 [[-0.2907075]
 [ 1.8342264]]
Final w1

Final w1 is: 
 [[-0.2015849]
 [ 1.8307562]]
Final w1 is: 
 [[-0.20101728]
 [ 1.830672  ]]
Final w1 is: 
 [[-0.2004371]
 [ 1.8306273]]
Final w1 is: 
 [[-0.19982305]
 [ 1.8306469 ]]
Final w1 is: 
 [[-0.19923471]
 [ 1.8305962 ]]
Final w1 is: 
 [[-0.19868825]
 [ 1.8305373 ]]
Final w1 is: 
 [[-0.19814038]
 [ 1.8304567 ]]
Final w1 is: 
 [[-0.1975665]
 [ 1.8304964]]
Final w1 is: 
 [[-0.1970097]
 [ 1.8304256]]
Final w1 is: 
 [[-0.1964817]
 [ 1.8303331]]
Final w1 is: 
 [[-0.1959846]
 [ 1.8301566]]
Final w1 is: 
 [[-0.19554305]
 [ 1.8299215 ]]
Final w1 is: 
 [[-0.19513412]
 [ 1.8296883 ]]
Final w1 is: 
 [[-0.19465843]
 [ 1.8295251 ]]
Final w1 is: 
 [[-0.19418135]
 [ 1.8293824 ]]
Final w1 is: 
 [[-0.19362807]
 [ 1.8293082 ]]
Final w1 is: 
 [[-0.1930872]
 [ 1.8292072]]
Final w1 is: 
 [[-0.1925231]
 [ 1.829111 ]]
Final w1 is: 
 [[-0.19194652]
 [ 1.8290542 ]]
Final w1 is: 
 [[-0.19133599]
 [ 1.8290625 ]]
Final w1 is: 
 [[-0.19075109]
 [ 1.8290002 ]]
Final w1 is: 
 [[-0.19020806]
 [ 1.8289299 ]]
Fina

 [[-0.0530821]
 [ 1.7783438]]
Final w1 is: 
 [[-0.05257823]
 [ 1.7781202 ]]
Final w1 is: 
 [[-0.05205209]
 [ 1.7780144 ]]
Final w1 is: 
 [[-0.05154173]
 [ 1.7777944 ]]
Final w1 is: 
 [[-0.05106265]
 [ 1.7775459 ]]
Final w1 is: 
 [[-0.05061518]
 [ 1.7772063 ]]
Final w1 is: 
 [[-0.05022095]
 [ 1.7768121 ]]
Final w1 is: 
 [[-0.04986005]
 [ 1.7764188 ]]
Final w1 is: 
 [[-0.04943259]
 [ 1.7760975 ]]
Final w1 is: 
 [[-0.04900062]
 [ 1.7758069 ]]
Final w1 is: 
 [[-0.04849169]
 [ 1.7755891 ]]
Final w1 is: 
 [[-0.04799615]
 [ 1.775341  ]]
Final w1 is: 
 [[-0.04747777]
 [ 1.7750965 ]]
Final w1 is: 
 [[-0.04694786]
 [ 1.7748919 ]]
Final w1 is: 
 [[-0.04638319]
 [ 1.7747587 ]]
Final w1 is: 
 [[-0.04584334]
 [ 1.7745546 ]]
Final w1 is: 
 [[-0.04534489]
 [ 1.7743434 ]]
Final w1 is: 
 [[-0.04484275]
 [ 1.7741148 ]]
Final w1 is: 
 [[-0.04431849]
 [ 1.7740041 ]]
Final w1 is: 
 [[-0.04380995]
 [ 1.773779  ]]
Final w1 is: 
 [[-0.04333279]
 [ 1.7735252 ]]
Final w1 is: 
 [[-0.04288725]
 [ 1.7731801 ]]
Fina

Final w1 is: 
 [[0.10514227]
 [1.6819333 ]]
Final w1 is: 
 [[0.10562185]
 [1.6816523 ]]
Final w1 is: 
 [[0.10606896]
 [1.6813395 ]]
Final w1 is: 
 [[0.10648483]
 [1.680936  ]]
Final w1 is: 
 [[0.10684967]
 [1.6804826 ]]
Final w1 is: 
 [[0.10718141]
 [1.6800296 ]]
Final w1 is: 
 [[0.10757872]
 [1.6796471 ]]
Final w1 is: 
 [[0.10798215]
 [1.6792994 ]]
Final w1 is: 
 [[0.10846157]
 [1.6790236 ]]
Final w1 is: 
 [[0.10892726]
 [1.6787175 ]]
Final w1 is: 
 [[0.10941511]
 [1.678414  ]]
Final w1 is: 
 [[0.10991348]
 [1.678148  ]]
Final w1 is: 
 [[0.11044607]
 [1.677954  ]]
Final w1 is: 
 [[0.11095428]
 [1.6776913 ]]
Final w1 is: 
 [[0.11142265]
 [1.6774236 ]]
Final w1 is: 
 [[0.11189578]
 [1.6771412 ]]
Final w1 is: 
 [[0.11238899]
 [1.6769707 ]]
Final w1 is: 
 [[0.1128673]
 [1.6766886]]
Final w1 is: 
 [[0.11331312]
 [1.6763744 ]]
Final w1 is: 
 [[0.11372776]
 [1.6759698 ]]
Final w1 is: 
 [[0.11409147]
 [1.6755157 ]]
Final w1 is: 
 [[0.11442212]
 [1.6750618 ]]
Final w1 is: 
 [[0.11481826]
 [1.6

Final w1 is: 
 [[0.19733123]
 [1.616723  ]]
Final w1 is: 
 [[0.19780877]
 [1.6165427 ]]
Final w1 is: 
 [[0.19827183]
 [1.6162542 ]]
Final w1 is: 
 [[0.19870229]
 [1.6159327 ]]
Final w1 is: 
 [[0.1991023]
 [1.6155235]]
Final w1 is: 
 [[0.19945288]
 [1.6150675 ]]
Final w1 is: 
 [[0.19977102]
 [1.6146116 ]]
Final w1 is: 
 [[0.20015348]
 [1.6142241 ]]
Final w1 is: 
 [[0.20054251]
 [1.6138719 ]]
Final w1 is: 
 [[0.20100594]
 [1.6135894 ]]
Final w1 is: 
 [[0.2014557]
 [1.6132776]]
Final w1 is: 
 [[0.20192695]
 [1.6129678 ]]
Final w1 is: 
 [[0.20240803]
 [1.6126934 ]]
Final w1 is: 
 [[0.20292233]
 [1.6124889 ]]
Final w1 is: 
 [[0.2034125]
 [1.6122181]]
Final w1 is: 
 [[0.20386444]
 [1.611944  ]]
Final w1 is: 
 [[0.20432141]
 [1.6116562 ]]
Final w1 is: 
 [[0.2047976]
 [1.6114756]]
Final w1 is: 
 [[0.20525934]
 [1.611187  ]]
Final w1 is: 
 [[0.20568849]
 [1.6108655 ]]
Final w1 is: 
 [[0.20608726]
 [1.6104565 ]]
Final w1 is: 
 [[0.20643675]
 [1.6100008 ]]
Final w1 is: 
 [[0.20675386]
 [1.6095454

 [[0.3422499]
 [1.509051 ]]
Final w1 is: 
 [[0.34261355]
 [1.5087134 ]]
Final w1 is: 
 [[0.34304753]
 [1.5084398 ]]
Final w1 is: 
 [[0.34346825]
 [1.5081396 ]]
Final w1 is: 
 [[0.34390897]
 [1.5078406 ]]
Final w1 is: 
 [[0.34435815]
 [1.5075723 ]]
Final w1 is: 
 [[0.3448381]
 [1.507369 ]]
Final w1 is: 
 [[0.3452945]
 [1.5071044]]
Final w1 is: 
 [[0.34571624]
 [1.5068394 ]]
Final w1 is: 
 [[0.3461432]
 [1.5065622]]
Final w1 is: 
 [[0.34658802]
 [1.5063834 ]]
Final w1 is: 
 [[0.34701926]
 [1.5061042 ]]
Final w1 is: 
 [[0.34741837]
 [1.505791  ]]
Final w1 is: 
 [[0.34778914]
 [1.5053971 ]]
Final w1 is: 
 [[0.34811404]
 [1.5049622 ]]
Final w1 is: 
 [[0.34840834]
 [1.504527  ]]
Final w1 is: 
 [[0.3487637]
 [1.5041553]]
Final w1 is: 
 [[0.349126 ]
 [1.5038188]]
Final w1 is: 
 [[0.3495584]
 [1.503546 ]]
Final w1 is: 
 [[0.3499776]
 [1.5032468]]
Final w1 is: 
 [[0.35041672]
 [1.5029486 ]]
Final w1 is: 
 [[0.35086423]
 [1.502681  ]]
Final w1 is: 
 [[0.35134235]
 [1.5024781 ]]
Final w1 is: 
 [[0

Final w1 is: 
 [[0.475454]
 [1.40834 ]]
Final w1 is: 
 [[0.47589144]
 [1.4081485 ]]
Final w1 is: 
 [[0.47630614]
 [1.4079013 ]]
Final w1 is: 
 [[0.47669083]
 [1.4076574 ]]
Final w1 is: 
 [[0.47708076]
 [1.4074028 ]]
Final w1 is: 
 [[0.47748712]
 [1.4072362 ]]
Final w1 is: 
 [[0.47788095]
 [1.4069785 ]]
Final w1 is: 
 [[0.47824362]
 [1.4066857 ]]
Final w1 is: 
 [[0.47858077]
 [1.4063209 ]]
Final w1 is: 
 [[0.47887647]
 [1.4059215 ]]
Final w1 is: 
 [[0.47914407]
 [1.405521  ]]
Final w1 is: 
 [[0.47946844]
 [1.4051787 ]]
Final w1 is: 
 [[0.4797999]
 [1.4048704]]
Final w1 is: 
 [[0.48019573]
 [1.4046191 ]]
Final w1 is: 
 [[0.48057908]
 [1.4043448 ]]
Final w1 is: 
 [[0.48098034]
 [1.4040705 ]]
Final w1 is: 
 [[0.4813883]
 [1.4038212]]
Final w1 is: 
 [[0.4818235]
 [1.4036305]]
Final w1 is: 
 [[0.48223603]
 [1.4033842 ]]
Final w1 is: 
 [[0.48261878]
 [1.4031415 ]]
Final w1 is: 
 [[0.48300678]
 [1.4028881 ]]
Final w1 is: 
 [[0.48341113]
 [1.4027222 ]]
Final w1 is: 
 [[0.48380303]
 [1.4024657 ]

Final w1 is: 
 [[0.54426163]
 [1.3557818 ]]
Final w1 is: 
 [[0.54464597]
 [1.3555448 ]]
Final w1 is: 
 [[0.5450555]
 [1.3553627]]
Final w1 is: 
 [[0.5454428]
 [1.3551283]]
Final w1 is: 
 [[0.5458033]
 [1.3548995]]
Final w1 is: 
 [[0.546169 ]
 [1.3546609]]
Final w1 is: 
 [[0.54655015]
 [1.3545041 ]]
Final w1 is: 
 [[0.5469195]
 [1.3542619]]
Final w1 is: 
 [[0.54725844]
 [1.353984  ]]
Final w1 is: 
 [[0.5475738]
 [1.3536396]]
Final w1 is: 
 [[0.5478506]
 [1.3532645]]
Final w1 is: 
 [[0.548101]
 [1.352888]]
Final w1 is: 
 [[0.5484053]
 [1.3525661]]
Final w1 is: 
 [[0.54871666]
 [1.3522776 ]]
Final w1 is: 
 [[0.5490886]
 [1.3520418]]
Final w1 is: 
 [[0.54944855]
 [1.3517852 ]]
Final w1 is: 
 [[0.5498251]
 [1.3515279]]
Final w1 is: 
 [[0.5502072]
 [1.3512921]]
Final w1 is: 
 [[0.5506143]
 [1.3511108]]
Final w1 is: 
 [[0.5509992]
 [1.3508776]]
Final w1 is: 
 [[0.55135757]
 [1.3506503 ]]
Final w1 is: 
 [[0.55172116]
 [1.3504131 ]]
Final w1 is: 
 [[0.5521001]
 [1.3502572]]
Final w1 is: 
 [[0.5

 [1.2648889]]
Final w1 is: 
 [[0.66300964]
 [1.2646859 ]]
Final w1 is: 
 [[0.663322 ]
 [1.2644664]]
Final w1 is: 
 [[0.6636481]
 [1.2642449]]
Final w1 is: 
 [[0.66397756]
 [1.2640378 ]]
Final w1 is: 
 [[0.6643272]
 [1.2638774]]
Final w1 is: 
 [[0.6646558]
 [1.2636719]]
Final w1 is: 
 [[0.66496444]
 [1.2634772 ]]
Final w1 is: 
 [[0.66527826]
 [1.2632742 ]]
Final w1 is: 
 [[0.6656056]
 [1.2631401]]
Final w1 is: 
 [[0.6659227]
 [1.2629324]]
Final w1 is: 
 [[0.6662109]
 [1.2626874]]
Final w1 is: 
 [[0.6664797]
 [1.2623876]]
Final w1 is: 
 [[0.6667163]
 [1.2620652]]
Final w1 is: 
 [[0.6669302]
 [1.2617407]]
Final w1 is: 
 [[0.66719174]
 [1.2614638 ]]
Final w1 is: 
 [[0.6674605]
 [1.2612188]]
Final w1 is: 
 [[0.6677814]
 [1.2610174]]
Final w1 is: 
 [[0.6680915]
 [1.2607998]]
Final w1 is: 
 [[0.66841525]
 [1.2605801 ]]
Final w1 is: 
 [[0.6687421]
 [1.2603744]]
Final w1 is: 
 [[0.66908896]
 [1.260215  ]]
Final w1 is: 
 [[0.6694148]
 [1.2600108]]
Final w1 is: 
 [[0.669721 ]
 [1.2598177]]
Final 

 [[0.72417593]
 [1.2179121 ]]
Final w1 is: 
 [[0.7244703]
 [1.2177925]]
Final w1 is: 
 [[0.72475535]
 [1.2176065 ]]
Final w1 is: 
 [[0.7250124]
 [1.2173822]]
Final w1 is: 
 [[0.72525275]
 [1.2171099 ]]
Final w1 is: 
 [[0.7254648]
 [1.2168201]]
Final w1 is: 
 [[0.7256565]
 [1.2165278]]
Final w1 is: 
 [[0.72589195]
 [1.2162788 ]]
Final w1 is: 
 [[0.72613466]
 [1.2160609 ]]
Final w1 is: 
 [[0.72642434]
 [1.2158811 ]]
Final w1 is: 
 [[0.7267039]
 [1.2156879]]
Final w1 is: 
 [[0.72699535]
 [1.2154917 ]]
Final w1 is: 
 [[0.7272884]
 [1.2153049]]
Final w1 is: 
 [[0.7275983]
 [1.2151593]]
Final w1 is: 
 [[0.72788787]
 [1.2149733 ]]
Final w1 is: 
 [[0.7281621]
 [1.2148017]]
Final w1 is: 
 [[0.72844154]
 [1.2146227 ]]
Final w1 is: 
 [[0.7287333]
 [1.2145044]]
Final w1 is: 
 [[0.72901577]
 [1.2143201 ]]
Final w1 is: 
 [[0.7292704]
 [1.2140973]]
Final w1 is: 
 [[0.7295085]
 [1.2138271]]
Final w1 is: 
 [[0.7297186]
 [1.2135398]]
Final w1 is: 
 [[0.7299086]
 [1.21325  ]]
Final w1 is: 
 [[0.730142 ]


 [[0.845468 ]
 [1.1239774]]
Final w1 is: 
 [[0.84564215]
 [1.1238326 ]]
Final w1 is: 
 [[0.8458494]
 [1.1237115]]
Final w1 is: 
 [[0.8460483]
 [1.1235845]]
Final w1 is: 
 [[0.8462542]
 [1.123452 ]]
Final w1 is: 
 [[0.8464574]
 [1.1233159]]
Final w1 is: 
 [[0.8466688]
 [1.123208 ]]
Final w1 is: 
 [[0.84686124]
 [1.1230706 ]]
Final w1 is: 
 [[0.84705067]
 [1.1229573 ]]
Final w1 is: 
 [[0.8472453]
 [1.1228392]]
Final w1 is: 
 [[0.84744936]
 [1.1227614 ]]
Final w1 is: 
 [[0.84764665]
 [1.1226358 ]]
Final w1 is: 
 [[0.8478182]
 [1.1224676]]
Final w1 is: 
 [[0.8479805]
 [1.1222708]]
Final w1 is: 
 [[0.84812546]
 [1.1220706 ]]
Final w1 is: 
 [[0.8482565]
 [1.1218662]]
Final w1 is: 
 [[0.84842104]
 [1.1216943 ]]
Final w1 is: 
 [[0.8485931]
 [1.1215518]]
Final w1 is: 
 [[0.84879786]
 [1.1214324 ]]
Final w1 is: 
 [[0.8489943]
 [1.1213075]]
Final w1 is: 
 [[0.8491976]
 [1.1211768]]
Final w1 is: 
 [[0.8493981]
 [1.1210424]]
Final w1 is: 
 [[0.84960645]
 [1.1209356 ]]
Final w1 is: 
 [[0.849796 ]
 [

Final w1 is: 
 [[0.8988181]
 [1.0829955]]
Final w1 is: 
 [[0.8989404]
 [1.0828586]]
Final w1 is: 
 [[0.899058 ]
 [1.0827042]]
Final w1 is: 
 [[0.8991647]
 [1.082555 ]]
Final w1 is: 
 [[0.8992612]
 [1.0824009]]
Final w1 is: 
 [[0.8993855]
 [1.0822731]]
Final w1 is: 
 [[0.89951783]
 [1.0821744 ]]
Final w1 is: 
 [[0.89967465]
 [1.0820903 ]]
Final w1 is: 
 [[0.89982414]
 [1.0820053 ]]
Final w1 is: 
 [[0.89997756]
 [1.0819126 ]]
Final w1 is: 
 [[0.90012527]
 [1.0818077 ]]
Final w1 is: 
 [[0.90027547]
 [1.0817229 ]]
Final w1 is: 
 [[0.90040725]
 [1.081615  ]]
Final w1 is: 
 [[0.9005443]
 [1.0815387]]
Final w1 is: 
 [[0.9006867]
 [1.0814588]]
Final w1 is: 
 [[0.9008369]
 [1.081407 ]]
Final w1 is: 
 [[0.9009818]
 [1.0813183]]
Final w1 is: 
 [[0.9011017]
 [1.0811827]]
Final w1 is: 
 [[0.90121716]
 [1.0810304 ]]
Final w1 is: 
 [[0.901322 ]
 [1.0808836]]
Final w1 is: 
 [[0.90141684]
 [1.0807319 ]]
Final w1 is: 
 [[0.90153927]
 [1.0806062 ]]
Final w1 is: 
 [[0.9016697]
 [1.0805095]]
Final w1 is: 


Final w1 is: 
 [[0.93800426]
 [1.0524775 ]]
Final w1 is: 
 [[0.93809456]
 [1.0524348 ]]
Final w1 is: 
 [[0.93819064]
 [1.0523899 ]]
Final w1 is: 
 [[0.93829316]
 [1.0523617 ]]
Final w1 is: 
 [[0.9383917]
 [1.0523062]]
Final w1 is: 
 [[0.9384652]
 [1.0521983]]
Final w1 is: 
 [[0.9385387]
 [1.0520844]]
Final w1 is: 
 [[0.9386079]
 [1.0519851]]
Final w1 is: 
 [[0.9386705]
 [1.0518795]]
Final w1 is: 
 [[0.938756 ]
 [1.0517949]]
Final w1 is: 
 [[0.9388501]
 [1.0517399]]
Final w1 is: 
 [[0.9389609]
 [1.0516907]]
Final w1 is: 
 [[0.93906516]
 [1.0516458 ]]
Final w1 is: 
 [[0.9391703]
 [1.0515908]]
Final w1 is: 
 [[0.93926644]
 [1.0515143 ]]
Final w1 is: 
 [[0.9393594]
 [1.0514503]]
Final w1 is: 
 [[0.9394338]
 [1.051369 ]]
Final w1 is: 
 [[0.93952215]
 [1.0513277 ]]
Final w1 is: 
 [[0.9396162]
 [1.0512842]]
Final w1 is: 
 [[0.93971664]
 [1.0512571 ]]
Final w1 is: 
 [[0.9398132]
 [1.051203 ]]
Final w1 is: 
 [[0.93988466]
 [1.0510962 ]]
Final w1 is: 
 [[0.93995637]
 [1.050984  ]]
Final w1 is: 


 [[0.9627369]
 [1.0333287]]
Final w1 is: 
 [[0.96280587]
 [1.0333031 ]]
Final w1 is: 
 [[0.9628626]
 [1.0332472]]
Final w1 is: 
 [[0.96291125]
 [1.0331981 ]]
Final w1 is: 
 [[0.9629404]
 [1.0331358]]
Final w1 is: 
 [[0.9629914]
 [1.0331224]]
Final w1 is: 
 [[0.9630488]
 [1.0331078]]
Final w1 is: 
 [[0.96311176]
 [1.0331006 ]]
Final w1 is: 
 [[0.9631718]
 [1.0330733]]
Final w1 is: 
 [[0.9632055]
 [1.0329862]]
Final w1 is: 
 [[0.9632435]
 [1.0329028]]
Final w1 is: 
 [[0.96328276]
 [1.032843  ]]
Final w1 is: 
 [[0.96331847]
 [1.0327754 ]]
Final w1 is: 
 [[0.96337366]
 [1.032725  ]]
Final w1 is: 
 [[0.9634386]
 [1.032706 ]]
Final w1 is: 
 [[0.963514 ]
 [1.0326855]]
Final w1 is: 
 [[0.96358323]
 [1.0326737 ]]
Final w1 is: 
 [[0.96365076]
 [1.0326495 ]]
Final w1 is: 
 [[0.9637059]
 [1.0325944]]
Final w1 is: 
 [[0.96375275]
 [1.0325459 ]]
Final w1 is: 
 [[0.96378  ]
 [1.0324843]]
Final w1 is: 
 [[0.96382946]
 [1.0324721 ]]
Final w1 is: 
 [[0.96388537]
 [1.0324587 ]]
Final w1 is: 
 [[0.9639467

- 第一种得到的 w1= [[1.019347 ],[1.0428089]],也就是得到的预测函数 $y = 1.02x_1 + 1.04x_2$ 比 $x_1 + x_2$ 大，因为在损失函数中，规定的是 预测少了损失更大，所以为了降低损失，预测值会更大些
- 第二种，将 loss_more 和 loss_less 调整后，w1 = [[0.9552581],[0.9813394]] ,预测多了损失更大，所以，少预测一些会降低损失
- 第三种，使用 MSE 均方误差作为损失函数 [[0.9743756],[1.0243336]] ，离标准答案更近


### 4.3 神经网络优化算法

本节将更加具体的介绍如何通过**反向传播算法（BP backpropgation）** 和**梯度下降算法(gradient descent)**调整神经网络中参数的取值。

- 梯度下降算法主要用于优化单个参数的取值
- 而反向传播算法给出了一个高效的方式在所有参数上使用梯度下降算法，从而使神经网络模型在训练数据上的损失函数尽可能小
- 略（略了很多，看原书）
- 只有当损失函数为凸函数时，梯度下降算法才能保证达到全局最优解
- GD : 计算时间太长，太慢
- SGD：为了加速训练过程，使用随机梯度下降的算法（stochastic gradient descent SGD）这个算法优化的不是在全部的训练数据上的损失函数，而是在每一轮迭代中，随机优化某一条训练数据上的损失函数。缺：在某一条数据上损失函数更小，不代表全部数据上的损失函数更小
- mini-batch GD :折中的方法，一小部分数据成为一个 batch，batch-size 是这一小部分的大小，可以大大减小收敛所需要的迭代次数


```

batch_size = n

# 每次读取一小部分数据作为当前数据来执行反向传播算法

x = tf.placeholder(tf.float32, shape=(batch_size, 2), name='x-input')
y_ = tf.placeholder(tf.float32, shape=(batch_size, 2), name='y-input')

# 定义神经网络结构和优化算法

loss = None
train_step = tf.train.AdamOptimizer(0.001).minimize(loss)

# 训练神经网络
with tf.Session() as sess:

    # 参数初始化
    None
    # 迭代的更新参数
    for i in range(STEPS):
        # 准备 batch_size 个训练数据。
        # 一般将所有的训练数据随机打乱之后再选取可以得到更好的优化效果
        current_X, current_Y = ...
        sess.run(train_step, feed_dict={x: current_X, y_:current_Y})
```


### 4.4 神经网络进一步优化

本节继续介绍神经网络优化过程中可能遇到的一些问题，以及解决这些问题的常用方法。

#### 4.4.1  学习率的设置

如何设置学习率（learning rate）？

通过指数衰减的方法来设置梯度下降算法中的学习率，既可以让模型在训练前期快速接近较优解，又可以保证模型在训练后期不会有太大的波动。

- 学习率不能过大，也不能过小，Tensorflow 提供了一种更加灵活的学习率设置方法——指数衰减法。
- tf.train.exponential_decay 函数实现了指数衰减学习率

- exponential_decay 函数会指数级地减小学习率，它实现以下代码的功能

```
decay_learning_rate = learning_rate & decay_rate ^ (global_step / decay_steps)

```

    - decay_learning_rate :每一轮优化时使用的学习率
    - learning_rate : 初始学习率
    - decay_rate ：衰减速度

- tf.train.exponential_decay 函数可以通过设置参数 staircase 选择不同的衰减方式，staircase= False ,平滑连续衰减 staircase= True 阶梯状衰减 

![](./images/decay.png)

代码示例：

#### 假设我们要最小化函数  $y=x^2$, 选择初始点   $x_0=5$

#### 1. 学习率为1的时候，x 在 5 和 -5 之间震荡。


In [ ]:
import tensorflow as tf

# 通过 exponential_decay 函数生成学习率

# 使用指数衰减的学习率。在 minimize 函数中传入 global_step 将自动更新
# global_step 参数，从而使得学习率也得到相应更新


#### 4.4.2 过拟合问题

#### 4.4.3 滑动平均模型